### DISCUSSION TITLE
Implementing a ProseMirror Remark plugin - How to work with non-unist, non-string inputs?

### DISCUSSION THREAD

**Discussion by smoores-dev** [unifiedjs/unified#254] [Q&A] [answered]
Howdy! I do a lot of work with ProseMirror, a rich text editing library. Recently I've been working with some folks that are interested in serializing ProseMirror documents to Markdown, and I love the unified/remark stack for working with Markdown.

I started by working out a basic remark plugin that converts mdast to a ProseMirror node tree (you can see that here: https://github.com/handlewithcarecollective/remark-prosemirror). I used the `CompileResultMap` interface to add a new result type for ProseMirror docs:

```ts
declare module "unified" {
  interface CompileResultMap {
    PmNode: PmNode;
  }
}
```

Which is awesome, I really appreciate how extensible this is. Now I'm trying to go in the opposite direction, from ProseMirror nodes to mdast/Markdown, and I'm not sure what the correct way to model that is.

What I _want_ to be able to do is just the inverse operation: tell unified to start with a ProseMirror node and output a markdown vFile. This doesn't really seem to align with unified's processing model, though. ProseMirror nodes have already been "parsed" from the DOM, so `processor.parse` (which takes only a string or byte array) doesn't really work or make sense, and `processor.run` takes a unist tree. ProseMirror docs are structured similarly to unist trees, but don't actually implement the interface.

It seems like the move here is to produce a unist-compatible ProseMirror syntax-tree, convert the ProseMirror doc to that tree, and then `processor.run` on the tree to convert it to mdast and stringify to Markdown.

Does that seem right? Or am I missing something about how to handle situations like this? It seems like comparable plugins like rehype-react are essentially one way; they don't have a way to get from the CompileResult back to on AST.

**Comment by ChristianMurphy** [comment, marked as answer]
> This doesn't really seem to align with unified's processing model, though

Unified has several different approaches to being used.
You appear to be referring mostly to the programmatic Node/JS API.
The string requirement comes in because it also has a CLI that expects to be able to read a file.

https://unifiedjs.com/learn/

---

> It seems like the move here is to produce a unist-compatible ProseMirror syntax-tree, convert the ProseMirror doc to that tree, and then processor.run on the tree to convert it to mdast and stringify to Markdown.

Exactly
I'd recommend creating an mdast util, there may be some inspiration to draw from a now deprecated quill connector https://github.com/syntax-tree/mdast-util-from-quill-delta

---

> ProseMirror nodes have already been "parsed" from the DOM, so processor.parse (which takes only a string or byte array) doesn't really work or make sense, and processor.run takes a unist tree. ProseMirror docs are structured similarly to unist trees, but don't actually implement the interface.

I'd recommend using `unified.run` if you choose to use the unified abstraction at all.
It is fine to leverage the mdast utils directly too.

**Reply by smoores-dev** [reply]
Thanks so much @ChristianMurphy! And yes, I'm referring to the programmatic JS API, that's right.

Thanks for sharing that Quill example! It looks like that takes a Quill data structure and turns it directly into an mdast tree, right? So if someone wanted to use that with unified, they would do something like:

```ts
const mdast = deltaToMdast(delta)

const markdown = unified().use(remarkStringify).run(mdast)
```

Is that right? Some silly part of my brain is upset that it's not symmetrical with the inverse operation:

```ts
// In here, remarkProseMirror is a new plugin based on an mdast util
// proseMirrorFromMdast
const proseMirrorNode = unified().use(remarkParse).use(remarkProseMirror).process(markdown)
```

But I suppose that doesn't actually matter haha.

**Reply by ChristianMurphy** [reply]
> Is that right?

It is

> Some silly part of my brain is upset that it's not symmetrical with the inverse operation

unified is mostly intended for string -> ast -> string.
the only exception I'm aware of right now is string -> ast -> hyperscript (react).
https://github.com/rehypejs/rehype-react

I'd be open to supporting the input be configurable, my main concern is the complexity that would add to the TypeScript types, which are already quite complex.

**Reply by smoores-dev** [reply]
Thanks again for your help. I just published [remark-prosemirror](https://www.npmjs.com/package/@handlewithcare/remark-prosemirror) last night, based on this conversation!

I think it could be neat if unified supported configurable input types. I imagine that Lexical and Slate folks might also be interested in interoperating with remark 🤷. 

I agree that the types are already pretty complicated (I actually had to cast through `unknown` in order to properly type the remark plugin, which... feels like I might be missing something but I don't know what??), and I'm sure typing everything with JSDoc is an additional challenge (but I appreciate the value of being able to directly ship the source code without a build step).

If it's something you'd be interested in accepting a PR for, I would be happy to look into it! It's not the end of the world, but the current situation means that consumers of remark-prosemirror that are less familiar with unified and the syntax-tree world need to understand not just the idea of remark plugins, but also become familiar with mdast (at least a bit).

**Reply by ChristianMurphy** [reply]
Thanks for sharing!
A few things, I see quickly looking through.
1. the repository doesn't seem to be recognized by npm https://www.npmjs.com/package/@handlewithcare/remark-prosemirror it would be good to have that so users can find the issue tracker, here are the docs on supported formats: https://docs.npmjs.com/cli/v10/configuring-npm/package-json?v=true#repository
2. SWC is listed as a production dependency, but only seems to be used by dev tools? Maybe it should be dev dependency? https://github.com/handlewithcarecollective/remark-prosemirror/blob/5e0a6ad3fde0f9fc7752347539a92b82f9aba83f/package.json#L66-L67
3. Where possible use a wide version range for dependencies https://github.com/handlewithcarecollective/remark-prosemirror/blob/5e0a6ad3fde0f9fc7752347539a92b82f9aba83f/package.json#L68-L74 For example `^4.0.0` is preferable to `^4.2.1` because it gives package managers more opportunities to de-duplicate.
4. It can be simplier to manage formatters together https://github.com/handlewithcarecollective/remark-prosemirror/blob/5e0a6ad3fde0f9fc7752347539a92b82f9aba83f/package.json#L27-L28 prettier can be run as an ESLint plugin https://github.com/prettier/eslint-config-prettier
5. There should need to be an unknown cast https://github.com/handlewithcarecollective/remark-prosemirror/blob/5e0a6ad3fde0f9fc7752347539a92b82f9aba83f/lib/remark-prosemirror.ts#L20 try following this approach for typing a plugin https://github.com/remcohaszing/remark-mdx-frontmatter/blob/6e33330fad5d252389a0aaa016bb9638b8d6f057/src/remark-mdx-frontmatter.ts#L35
6. Is it intentional there is a version of yarn bundled in the git repo? https://github.com/handlewithcarecollective/remark-prosemirror/tree/main/.yarn/releases
7. If you intend this plugin to be used with TypeScript and with Yarn, you may want/need to move these to dependencies so they can be resolved by pnpm and newer versions of yarn https://github.com/handlewithcarecollective/remark-prosemirror/blob/5e0a6ad3fde0f9fc7752347539a92b82f9aba83f/package.json#L41-L42

**Reply by smoores-dev** [reply]
Thanks for the review @ChristianMurphy, I really appreciate it!!

1 - 3, great points, I will resolve!

4 - That's a link to the prettier _config_ for eslint, which only turns off formatting rules in eslint that conflict with prettier. As I understand it, that's no longer necessary with eslint v9, which disables those rules by default. The choice to separate linting and formatting was intentional, as the prettier docs [don't recommend using the eslint plugin](https://prettier.io/docs/en/integrating-with-linters.html#notes)! It's slower and more complicated than just running prettier on its own.

5 - thanks for that! I feel like it's closer now, but something's still wonky... If I type the plugin as `Plugin<[Options], MdastRoot, PmNode>`, then `this.compiler` is typed as a function that receives a `PmNode` as an argument. Shouldn't it be an `MdastRoot`?

6 - yes! I think Yarn recently started recommending switching to corepack for this, but prior to that the preferred installation method was to actually check the yarn executable into your repo: https://yarnpkg.com/configuration/yarnrc#yarnPath. I'll take a look at switching to corepack (it's just kind of a pain to get working with Nix haha)

7 - ... huh! This is a good point, I hadn't really thought about it. I'll play around with it, but this makes sense as probably being necessary

**Reply by ChristianMurphy** [reply]
> as the prettier docs [don't recommend using the eslint plugin](https://prettier.io/docs/en/integrating-with-linters.html#notes)!

I'm aware, and disagree 😅 
The complexity of running several tools more than having one config, and having all the rules in eslint I've found creates less conficts when some eslint plugins do some stylistic rules as well.

> I'll take a look at switching to corepack (it's just kind of a pain to get working with Nix haha)

Hopefully corepack gets smoother with nix.
I'll also note that, with the most recent versions of `npm` they are often faster and just as feature rich as `yarn`

**Reply by smoores-dev** [reply]
Re: the plugin type -- part of the issue seems to be that the Plugin type doesn't actually pass its type parameters through to the Processor type. So it doesn't matter that I'm saying this is a Plugin that takes an MdastRoot and compiles to a PmNode, because it's then typed as a Processor that takes a unist Node, because that's the default type value. I suspect that this is the actual issue -- separately, it seems like the Typescript integration in VS Code has a bug when there are naming conflicts like this, and thinks that the Node type that compile takes as an arg is the prosemirror-model Node, but it's actually the unist Node, and it's linking me to the wrong place!

**Reply by smoores-dev** [reply]
Yup, confirmed that it's _not_ a PmNode, it's a unist node, and VS Code/Typescript/tsserver is just linking me to the wrong Node type. This is better, but still has a cast:

```ts
export const remarkProseMirror: Plugin<[Options], MdastRoot, PmNode> =
  function (options: Options) {
    this.compiler = function (tree) {
      return toProseMirror(tree as MdastRoot, options);
    };
  };
```

**Reply by wooorm** [reply]
Options are perhaps optional, so that would be: `Plugin<[(Options | null | undefined)?], MdastRoot, PmNode>` and `(options?: Options | null | undefined)`.

To type your custom outputs, register it:

https://github.com/unifiedjs/unified/blob/6838c70bd2b237663f522a0b46fccf7aef8491bc/lib/index.js#L13

See e.g. https://github.com/rehypejs/rehype-react/blob/93fac074e8e3447088ed2408282e9e089ea7b36c/index.d.ts#L42
